### PR TITLE
Feature/define iobject

### PR DESCRIPTION
**Pull request by omskscream**
Developed interface IObject with explicit javadocs,
Developed tests for all IObject implementations with runtime classes-resolving,
Developed helpers for reflection operations and base exception class for IObject methods.

### CODE REVIEW

**Comment by omskscream** [comment]
Work is not done yet.

**Comment by sharipov-ru** [comment]
@omskscream I suggest you to introduce a special GitHub label for marking pull request which are not done yet. For example, label `wip` (work in progress)

**Comment by omskscream** [comment]
Label added, pull request reopened.

**Comment by omskscream** [comment]
Should close #1;
should close #2.

File: src/main/java/info/smart_tools/smartactors/core/CoreException.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package info.smart_tools.smartactors.core;
+
+/**
+ * General exception of smartactors core.
+ * Other core exceptions should extend {@code CoreException}.
+ */
+public class CoreException extends Exception {
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/CoreException.java]: Why did you introduce this exception? What does it mean?

File: src/main/java/info/smart_tools/smartactors/core/IObject.java [hunk 2]
```diff
@@ -1,26 +1,47 @@
...
+     *             - \ r
+     *             - \ t
+     *             - \ u four-hex-digits
+     *             or be empty string
+     * @param newValue is new value of field, it can be any Java object or {@code null}
+     * @throws InvalidNameException if name argument is invalid
+     * @throws BaseIObjectException in case of underlying exceptions
      */
-    void setValue(String name, Object newValue);
+    void setValue(String name, Object newValue) throws BaseIObjectException;
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/IObject.java]: Please use ChangeValueException name for this exception

File: src/main/java/info/smart_tools/smartactors/core/IObject.java [hunk 3]
```diff
@@ -1,26 +1,47 @@
...
+     *             - \ t
+     *             - \ u four-hex-digits
+     *             or be empty string
      * @return named field value object
      *         or {@code null} if field does not exist
+     * @throws InvalidNameException if name argument is invalid
+     * @throws BaseIObjectException in case of underlying exceptions
      */
-    Object getValue(String name);
+    Object getValue(String name) throws BaseIObjectException;
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/IObject.java]: Please use ReadValueException name for this case.

File: src/main/java/info/smart_tools/smartactors/core/BaseIObjectException.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * {@code IObject} interface to indicate that concrete
+ * implementation received an exception from its underlying operations.
+ *
+ * Developer should extend {@code BaseIObjectException} to more concrete
+ * for implementation of IObject, if needed (e.g. database access or file I/O).
+ *
+ * @see IObject
+ * @see CoreException
+ */
+public class BaseIObjectException extends CoreException {
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/BaseIObjectException.java]: Please remove this exception

File: src/main/java/info/smart_tools/smartactors/core/CreationException.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package info.smart_tools.smartactors.core;
+
+/**
+ * The {@code CreationException} should be thrown if {@code IObject}
+ * implementation instance cannot be created for any reason.
+ */
+public class CreationException extends BaseIObjectException {
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/CreationException.java]: Please remove this exception

File: src/main/java/info/smart_tools/smartactors/core/IObjectCreator.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ * IObjectCreator knows how to create IObject objects and can create them.
+ *
+ * Every {@code IObject} implementation should have its own
+ * creator class that implements this interface.
+
+ * Each {@code IObjectCreator} implementation must have a constructor without parameters
+ * which returns a fully initialized object with default settings.
+ *
+ * E.g. for some {@code IObject} implementation like
+ * {@code public class MyObject implements IObject {...}}
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/IObjectCreator.java]: What does MyObject name mean?

File: src/main/java/info/smart_tools/smartactors/core/InvalidNameException.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package info.smart_tools.smartactors.core;
+
+/**
+ * An {@code InvalidNameException} is thrown by methods of the
+ * {@code IObject} interface to indicate that given name of field
+ * is not valid field name.
+ *
+ * @see IObject
+ */
+public class InvalidNameException extends BaseIObjectException {
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/InvalidNameException.java]: I think in java there is a standard exception for this purpose.

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Default implementation of IObject interface.
+ */
+public class SMObject implements IObject {
+
+    private ObjectMapper objectMapper; //saved for later conversion to json
+    private Map<String, Object> map;
+
+    public SMObject() {
+        this.objectMapper = new ObjectMapper();
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Do you really need objectMapper this?

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        this.map = new HashMap<>(0);
+    }
+
+    public SMObject(String raw) throws CreationException {
+        this.objectMapper = new ObjectMapper();
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: I don't see a reason to have the objectMapper instance more than constuctor is working.

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        try {
+            this.map = this.objectMapper.readValue(raw, new TypeReference<Map<String, Object>>() {});
+        } catch (IOException e) {
+            throw new CreationException(e);
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: May be java has a suitable default exception for this case?

**Pull request merged**